runtime/src/serde_snapshot.rs

```diff
@@ -74,6 +74,62 @@ struct AccountsDbFields<T>(
     BankHashInfo,
 );
 
+/// Helper type to wrap BufReader streams when deserializing and reconstructing from either just a
+/// full snapshot, or both a full and incremental snapshot
+pub struct SnapshotStreams<'a, R> {
+    pub full_snapshot_stream: &'a mut BufReader<R>,
+    pub incremental_snapshot_stream: Option<&'a mut BufReader<R>>,
+}
+
+/// Helper type to wrap AccountsDbFields when reconstructing AccountsDb from either just a full
+/// snapshot, or both a full and incremental snapshot
+#[derive(Debug)]
+struct SnapshotAccountsDbFields<T> {
+    full_snapshot_accounts_db_fields: AccountsDbFields<T>,
+    incremental_snapshot_accounts_db_fields: Option<AccountsDbFields<T>>,
+}
+
+impl<T> SnapshotAccountsDbFields<T> {
+    /// Collapse the SnapshotAccountsDbFields into a single AccountsDbFields.  If there is no
+    /// incremental snapshot, this returns the AccountsDbFields from the full snapshot.  Otherwise
+    /// this uses the version, slot, and bank hash info from the incremental snapshot, then the
+    /// combination of the storages from both the full and incremental snapshots.
+    fn collapse_into(self) -> Result<AccountsDbFields<T>, Error> {
+        match self.incremental_snapshot_accounts_db_fields {
+            None => Ok(self.full_snapshot_accounts_db_fields),
+            Some(AccountsDbFields(
+                mut incremental_snapshot_storages,
+                incremental_snapshot_version,
+                incremental_snapshot_slot,
+                incremental_snapshot_bank_hash_info,
+            )) => {
+                let full_snapshot_storages = self.full_snapshot_accounts_db_fields.0;
+                let full_snapshot_slot = self.full_snapshot_accounts_db_fields.2;
```

Comment:
heh, really wish we'd gone with named fields for `AccountsDbFields` 😛

Reply:
Me too. I had grabbed `.1` for `slot` and all my tests were failing until I realized...